### PR TITLE
Split Terraform and Ruby CI actions into separate jobs.

### PR DESCRIPTION
**Pull request by sengi**
This speeds up the CI runs significantly by:

- Allowing the Ruby lint and tests to run in parallel with the Terraform lint/validation.
- Skipping the Ruby checks if no Ruby-related stuff has changed; similarly for Terraform.
- Only running `bundle install` once per run instead of twice.
- Only running `terraform init` once per root module per run instead of twice.
- Only fetching the latest revision from Git instead of all the history.

Also:

- Enables linting and validation for some Terraform modules which were missing checks before.
- Checks all Terraform root modules in `/terraform/deployments` without requiring us to add a magic `.tflint.hcl` config file when we create a new one.
- Fixes the problem where the Terraform checks abort at the module which has an error. Now continues to check the rest of the modules so you don't have to keep re-running the job.
- Uses terraform-linters/setup-tflint instead of rolling our own. This gives us [Problem Matchers] so that lint issues show up inline on PRs.
- Fixes the problem where a failed tflint install produces no useful error message.

Tested: Pushed a Ruby commit; Ruby checks ran. Pushed a Terraform commit; Terraform checks ran. Terraform checks continue to check all root modules rather than stopping at the first module which has an error.

[Problem Matchers]: https://github.com/actions/toolkit/blob/main/docs/problem-matchers.md

### CODE REVIEW

File: .github/workflows/ci-ruby.yml [hunk 1]
```diff
@@ -0,0 +1,13 @@
+name: Ruby linting and tests
+on:
+  push:
+    paths: ['**/*.rb', '**/*.rake', '.github/workflows/ci-ruby.yml']
```
Comment (bilbof) [inline .github/workflows/ci-ruby.yml:4]: Looks good. There are a few more Ruby files to add: `.ruby-version`, `.rspec`, `.rubocop.yml`, `Rakefile`, `Gemfile`.

Comment (sengi) [reply .github/workflows/ci-ruby.yml:4]: Aha, thanks

File: .github/workflows/ci-terraform.yml [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    paths: ['**/*.tf', '**/*.hcl', '.github/workflows/ci-terraform.yml']
+env:
+  TF_IN_AUTOMATION: true
+  TF_PLUGIN_CACHE_DIR: ${{ github.workspace }}/.terraform.d/plugin-cache
+jobs:
+  tflint:
+    runs-on: ubuntu-latest
+    steps:
+    - uses: actions/checkout@v2
+      name: Checkout source code
```
Comment (fredericfran-gds) [inline .github/workflows/ci-terraform.yml:13]: maybe checkout the latest commit only, i.e.:
with:
          fetch-depth: 0

File: .github/workflows/ci-ruby.yml (hunk 1 above)
Comment (bilbof) [reply .github/workflows/ci-ruby.yml:4]: I'm happy if we want to push this stuff down into a subdirectory (hard to name though, `scripts` or `ruby` both feel too generic, though we have directories for `docker` and `terraform`)

File: .github/workflows/ci-terraform.yml (hunk 2 above)
Comment (sengi) [reply .github/workflows/ci-terraform.yml:13]: That actually causes it to fetch all revisions. The default is `fetch-depth: 1`, which fetches only the latest revision. See https://github.com/actions/checkout#checkout-v2

Comment (fredericfran-gds) [reply .github/workflows/ci-terraform.yml:13]: thanks, I got confused on that one.

File: .github/workflows/ci-ruby.yml (hunk 1 above)
Comment (sengi) [reply .github/workflows/ci-ruby.yml]: Yeah, I wondered about moving stuff around but thought it's probably more robust to have the job not care where the source files are.

**Review by bilbof — APPROVED** [review]
Thanks for doing this! 🚀 

Looks like a reasonable change. It's nice having these as multiple actions, it's much easier to understand the various jobs and steps.

According to the GitHub docs if we break up the steps into separate `jobs` they will [run in parallel](https://docs.github.com/en/actions/learn-github-actions/understanding-github-actions) (e.g. you can run `lint` and `test` concurrently). Steps seem to run serially. Not necessary for this PR though!

**Comment by sengi** [comment]
> if we break up the steps into separate jobs they will run in parallel (e.g. you can run lint and test concurrently).

That's a good idea - I'll add that to my rainy-day TODO list. I think Jobs run in separate containers though so to share stuff you have to use caches. I'm pretty new to GH Actions though so I'm probably not making the best use of it yet.

**Pull request merged**